### ISSUE TITLE
[rllib] Entropy loss calculation in vtrace policy with autoregressive actions does not calculate entropy based on sample conditioned action distribution.

### ISSUE THREAD

**Issue by Bam4d** [ray-project/ray#16081]
### What is the problem?

rllib version: 2.0.0.dev
framework: torch + tf

### Reproduction

There is no *physical replication* as its a bug in calculation, but I will show and explain the code where the bug is present (in the *torch* framework).

Entropy loss for vtrace is calculated here:
https://github.com/ray-project/ray/blob/master/rllib/agents/impala/vtrace_torch_policy.py#L105

```
 # The entropy loss.        
self.entropy = torch.sum(actions_entropy * valid_mask)        
self.mean_entropy = self.entropy / torch.sum(valid_mask)
```

Assuming the autoregressive action has two parts, this entropy loss should be the entropy of p(a2|a1, s)+p(a1|s) where a1 and a2 are the two parts of the action. This action distribution *should* be the same action distribution calculated from the trjectory given by the current sample that is being processed. 

However what actually happens is entropy is calculated, not based on the sample trajectory, but from a random *re-sampled* action:

The entropy is calculated in the call to this function:
https://github.com/ray-project/ray/blob/master/rllib/agents/impala/vtrace_torch_policy.py#L159

E.g:
```
actions_logp=_make_time_major(            
action_dist.logp(actions), drop_last=True),        
actions_entropy=_make_time_major(            
action_dist.entropy(), drop_last=True),
```

The entropy calculation for autoregressive actions can be found here:
https://github.com/ray-project/ray/blob/d8f58e683f6ab68a36fe3a7e6d4b1ca91252e239/rllib/examples/models/autoregressive_action_dist.py#L119

```
 def entropy(self):
        a1_dist = self._a1_distribution()
        a2_dist = self._a2_distribution(a1_dist.sample())
        return a1_dist.entropy() + a2_dist.entropy()
```

In the above code, the entropy calculation is correct for the a1 (as the distribution is not dependent on the sampling) *BUT* the entropy for a2 is based on a *random sample* from a1. Which makes the entropy calculation for a2 incorrect.

N.B this is not the case for logp(actions) as the actions from the sample are used to calculate the correct distribution using the model.

The entropy function *could* be fixed by using the previously sampled action from the distirbution as is done in the `logp(actions) ` function, but its possible this will break many other algorithms.
```
 def entropy(self, actions):
        a1_dist = self._a1_distribution()
        a2 = actions[:, 1]
        a2_dist = self._a2_distribution(a2)
        return a1_dist.entropy() + a2_dist.entropy()
```

This issue only occurs in auto-regressive action spaces due to the sampling. This issue might occur in other methods where the entropy is being calculated, but I don't know where these might be.

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

**Comment by yangysc** [comment]
I also found this problem. But the true entropy is also sampled. Is a new resample operation for a2 action is not an approximation for the entropy, too?
